File: src/main/java/com/eager/questioncloud/resolver/QuestionFilterArgumentResolver.java
package com.eager.questioncloud.resolver;

import com.eager.questioncloud.question.dto.QuestionFilter;
import com.eager.questioncloud.question.model.QuestionLevel;
import com.eager.questioncloud.question.model.QuestionSortType;
import com.eager.questioncloud.question.model.QuestionType;
import com.eager.questioncloud.security.UserPrincipal;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.core.MethodParameter;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableHandlerMethodArgumentResolver;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

@Component
@RequiredArgsConstructor
public class QuestionFilterArgumentResolver implements HandlerMethodArgumentResolver {

    private final PageableHandlerMethodArgumentResolver pageableResolver;

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.getParameterType().equals(QuestionFilter.class);
    }

    @Override
    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest,
        WebDataBinderFactory binderFactory) throws Exception {
        Long userId = getUserIdFromRequest();
        List<Long> categories = getCategoriesFromRequest(webRequest);
        List<QuestionLevel> levels = getLevelsFromRequest(webRequest);
        QuestionType questionType = getQuestionTypeFromRequest(webRequest);
        Long creatorId = getCreatorIdFromRequest(webRequest);
        QuestionSortType sort = getSortFromRequest(webRequest);
        Pageable pageable = pageableResolver.resolveArgument(parameter, mavContainer, webRequest, binderFactory);
        return new QuestionFilter(userId, categories, levels, questionType, creatorId, sort, pageable);
    }

    private Long getUserIdFromRequest() {
        UserPrincipal userPrincipal = (UserPrincipal) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        return userPrincipal.getUser().getUid();
    }

    private List<Long> getCategoriesFromRequest(NativeWebRequest webRequest) {
        String input = webRequest.getParameter("categories");

        if (input == null) {
            return null;
        }

        String[] categoriesParams = input.split(",");

        return Arrays.stream(categoriesParams)
            .map(Long::parseLong)
            .collect(Collectors.toList());
    }

    private List<QuestionLevel> getLevelsFromRequest(NativeWebRequest webRequest) {
        String input = webRequest.getParameter("levels");

        if (input == null) {
            return null;
        }

        String[] levelParams = input.split(",");

        return Arrays.stream(levelParams)
            .map(QuestionLevel::valueOf)
            .collect(Collectors.toList());
    }

    private QuestionType getQuestionTypeFromRequest(NativeWebRequest webRequest) {
        String input = webRequest.getParameter("questionType");

        if (input == null) {
            return null;
        }

        return QuestionType.valueOf(input);
    }

    private Long getCreatorIdFromRequest(NativeWebRequest webRequest) {
        String input = webRequest.getParameter("creatorId");

        if (input == null) {
            return null;
        }

        return Long.parseLong(input);
    }

    private QuestionSortType getSortFromRequest(NativeWebRequest webRequest) {
        String input = webRequest.getParameter("sort");

        if (input == null) {
            return null;
        }

        return QuestionSortType.valueOf(input);
    }
}

File: qc-domain/qc-question/qc-question-api/src/main/kotlin/com/eager/questioncloud/question/resolver/QuestionFilterArgumentResolver.kt
package com.eager.questioncloud.question.resolver

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.question.common.QuestionFilter
import com.eager.questioncloud.question.enums.QuestionLevel
import com.eager.questioncloud.question.enums.QuestionSortType
import com.eager.questioncloud.question.enums.QuestionType
import org.springframework.core.MethodParameter
import org.springframework.stereotype.Component
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer
import java.util.*
import java.util.stream.Collectors

@Component
class QuestionFilterArgumentResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return parameter.parameterType == QuestionFilter::class.java
    }

    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?
    ): QuestionFilter {
        val categories = getCategoriesFromRequest(webRequest)
        val levels = getLevelsFromRequest(webRequest)
        val questionType = getQuestionTypeFromRequest(webRequest)
        val creatorId = getCreatorIdFromRequest(webRequest)
        val sort = getSortFromRequest(webRequest)
        return QuestionFilter(categories, levels, questionType, creatorId, sort)
    }

    private fun getCategoriesFromRequest(webRequest: NativeWebRequest): List<Long>? {
        val input = webRequest.getParameter("categories") ?: return null

        val categoriesParams = input.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        return Arrays.stream(categoriesParams)
            .map { s: String -> s.toLong() }
            .collect(Collectors.toList())
    }

    private fun getLevelsFromRequest(webRequest: NativeWebRequest): List<QuestionLevel>? {
        val input = webRequest.getParameter("levels") ?: return null

        val levelParams = input.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        return Arrays.stream(levelParams)
            .map { value: String? -> QuestionLevel.valueOf(value!!) }
            .collect(Collectors.toList())
    }

    private fun getQuestionTypeFromRequest(webRequest: NativeWebRequest): QuestionType? {
        val input = webRequest.getParameter("questionType") ?: return null

        return QuestionType.valueOf(input)
    }

    private fun getCreatorIdFromRequest(webRequest: NativeWebRequest): Long? {
        val input = webRequest.getParameter("creatorId") ?: return null

        return input.toLong()
    }

    private fun getSortFromRequest(webRequest: NativeWebRequest): QuestionSortType {
        val input = webRequest.getParameter("sort")

        if (input.isNullOrBlank()) {
            throw CoreException(Error.BAD_REQUEST)
        }

        return QuestionSortType.valueOf(input)
    }
}
